views/audit/big_query_elt_script_logging.sql

```diff
@@ -0,0 +1,525 @@
+      '$.jobChange.job.jobConfig.tableCopyConfig.sourceTables'),
+      '["'),'"]') ,'","') AS tableCopySourceTables,
+    CAST(JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
+      '$.jobChange.job.jobConfig.tableCopyConfig.sourceTablesTruncated') 
+      AS BOOL) AS tableCopyConfigSourceTablesTruncated,
+    JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
+      '$.jobChange.job.jobConfig.tableCopyConfig.destinationTable') AS tableCopyConfigDestinationTable,
+    JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
+      '$.jobChange.job.jobConfig.createDisposition') AS tableCopyConfigCreateDisposition,
+    JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
+      '$.jobChange.job.jobConfig.writeDisposition') AS tableCopyConfigWriteDisposition,
+    JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
+      '$.jobChange.job.jobConfig.tableCopyConfig.destinationTableEncryption.kmsKeyName') AS tableCopyConfigDestinationTableEncryptionKmsKeyName,
+    SPLIT(JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
+      '$.jobChange.job.jobConfig.tableCopyConfig.destinationTable'),
+      ".")[SAFE_OFFSET(1)] AS tableCopyConfigProjectId,
+    SPLIT(JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
+      '$.jobChange.job.jobConfig.tableCopyConfig.destinationTable'),
+      ".")[SAFE_OFFSET(2)]  AS tableCopyConfigDatasetId,
+    SPLIT(JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
+      '$.jobChange.job.jobConfig.tableCopyConfig.destinationTable'),
+      ".")[SAFE_OFFSET(3)] AS tableCopyConfigTableId,
+    JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson, '$.jobChange.before') AS jobChangeBefore,
+    JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson, '$.jobChange.after') AS jobChangeAfter,
+    REGEXP_EXTRACT(protopayload_auditlog.metadataJson, r'BigQueryAuditMetadata","(.*?)":') AS eventName,
+  FROM `project_id.dataset_id.cloudaudit_googleapis_com_data_access`
+),
+/*
+ * TableDeletion: Table deletion event
+ * https://cloud.google.com/bigquery/docs/reference/auditlogs/rest/Shared.Types/BigQueryAuditMetadata#tabledeletion
+ */
+tableDeletionEvent AS (
+  SELECT
+    CONCAT(
+      SPLIT(
+        JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson, '$.tableDeletion.jobName'),
+        "/")[SAFE_OFFSET(1)],
+      ":",
+      SPLIT(
+        JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson, '$.tableDeletion.jobName'),
+        "/")[SAFE_OFFSET(3)]
+      ) AS jobId,
+    JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson, '$.tableDeletion.jobName') AS tableDeletionJobName,
+    JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
+      '$.tableDeletion.table.reason') AS tableDeletionReason,
+  FROM `project_id.dataset_id.cloudaudit_googleapis_com_system_event`
+),
+/*
+ * TableDataRead: Data from tableDataRead audit logs
+ * https://cloud.google.com/bigquery/docs/reference/auditlogs/rest/Shared.Types/BigQueryAuditMetadata#tabledataread
+ */
+tableDataReadEvent AS (
```

Comment:
Turns out tableDataReadEvents have an row entry for each table that a job reads. We have to tweak this and any other metadata events that are similar. Below shows an example of pulling all rows into one. Only drawback of the below format is that tableDataReadFields are kept as JSON STRING arrays

/*
 * TableDataRead: Data from tableDataRead audit logs
 * https://cloud.google.com/bigquery/docs/reference/auditlogs/rest/Shared.Types/BigQueryAuditMetadata#tabledataread
 */
tableDataReadEvent AS (
  SELECT
    CONCAT(
      SPLIT(
        JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
            '$.tableDataRead.jobName'),
            "/")[SAFE_OFFSET(1)],
      ":",
      SPLIT(
        JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
          '$.tableDataRead.jobName'),
            "/")[SAFE_OFFSET(3)]
    ) AS jobId,
    ARRAY_AGG(protopayload_auditlog.resourceName
      IGNORE NULLS ORDER BY protopayload_auditlog.resourceName) AS tableName,
    ARRAY_AGG(JSON_EXTRACT(protopayload_auditlog.metadataJson, '$.tableDataRead.fields')
      IGNORE NULLS ORDER BY protopayload_auditlog.resourceName) AS tableDataReadFields,
    ARRAY_AGG(CAST(JSON_EXTRACT(protopayload_auditlog.metadataJson, '$.tableDataRead.fieldsTruncated') AS BOOL)
      IGNORE NULLS ORDER BY protopayload_auditlog.resourceName) AS tableDataReadFieldsTruncated,
    ARRAY_AGG(JSON_EXTRACT(protopayload_auditlog.metadataJson, '$.tableDataRead.categories')
      IGNORE NULLS ORDER BY protopayload_auditlog.resourceName) AS tableDataReadCategories,
    ARRAY_AGG(CAST(JSON_EXTRACT(protopayload_auditlog.metadataJson, '$.tableDataRead.categoriesTruncated') AS BOOL)
      IGNORE NULLS ORDER BY protopayload_auditlog.resourceName) AS tableDataReadCategoriesTruncated,
    ARRAY_AGG(JSON_EXTRACT(protopayload_auditlog.metadataJson, '$.tableDataRead.reason') 
      IGNORE NULLS ORDER BY protopayload_auditlog.resourceName) AS tableDataReadReason,
    ARRAY_AGG(JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson, '$.tableDataRead.jobName')
      IGNORE NULLS ORDER BY protopayload_auditlog.resourceName) AS tableDataReadJobName,
    ARRAY_AGG(JSON_EXTRACT(protopayload_auditlog.metadataJson, '$.tableDataRead.sessionName')
      IGNORE NULLS ORDER BY protopayload_auditlog.resourceName) AS tableDataReadSessionName,
  FROM `simba-jackal-poc.scripting_logs.cloudaudit_googleapis_com_data_access`
  GROUP BY 1
),

Reply:
In the example you gave, we don't need "ARRAY_AGG(JSON_EXTRACT_SCALAR(protopayload_auditlog.metadataJson,
'$.tableDataRead.jobName') IGNORE NULLS ORDER BY protopayload_auditlog.resourceName) AS tableDataReadJobName" then right?